domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresConnectionManager.java

```diff
@@ -0,0 +1,96 @@
+import org.apache.logging.log4j.Logger;
+import org.folio.rest.tools.utils.Envs;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PostgresConnectionManager {
+  private static final Logger LOG = LogManager.getLogger(PostgresConnectionManager.class);
+  private static final int DEFAULT_CACHE_SIZE = 4;
+  private final int cacheSizeLimit;
+  private final List<PgConnection> connectionCache = new ArrayList<>();
+  private String currentTenant = "";
+  private int currentIndex;
+
+  public PostgresConnectionManager () {
+    cacheSizeLimit = Envs.getEnv(Envs.DB_MAXSHAREDPOOLSIZE) != null ?
+        Integer.parseInt(Envs.getEnv(Envs.DB_MAXSHAREDPOOLSIZE)) : DEFAULT_CACHE_SIZE;
+  }
+
+  public void clearCache() {
+    currentTenant = "";
+    currentIndex = 0;
+    connectionCache.clear();
+
+    LOG.debug("Cleared connection manager");
+  }
+
+  public Future<PgConnection> getConnection(Pool pool, String schemaName, String tenantId) {
+    if (!PostgresClient.sharedPgPool) {
+      LOG.debug("Not in shared pool mode");
+      return pool.getConnection().map(PgConnection.class::cast);
+    }
+
+    return getCachedConnection(pool, schemaName, tenantId);
+  }
+
+  private Future<PgConnection> getCachedConnection(Pool pool, String schemaName, String tenantId) {
+    if (tenantId.equals(currentTenant)) {
+      var connection = tryGetCachedConnection();
+      if (connection != null) {
+        return Future.succeededFuture(connection);
+      }
+
+      LOG.debug("Adding to connection cache for tenant {} {}", tenantId, connectionCache.size());
+      return createAndCacheConnection(pool, schemaName, tenantId);
+    }
+
+    clearCache();
+
+    currentTenant = tenantId;
+
+    LOG.debug("Starting new connection cache for tenant {} {}", tenantId, connectionCache.size());
+    return createAndCacheConnection(pool, schemaName, tenantId);
+  }
+
+  private PgConnection tryGetCachedConnection() {
+    if (connectionCache.size() > cacheSizeLimit) {
+      return getCachedConnection();
+    }
+    return null;
+  }
+
+  private Future<PgConnection> createAndCacheConnection(Pool pool, String schemaName, String tenantId) {
+    return pool.getConnection().compose(sqlConnection -> {
+      String sql = PostgresClient.DEFAULT_SCHEMA.equals(tenantId)
+          ? "SET ROLE NONE; SET SCHEMA ''"
+          : ("SET ROLE '" + schemaName + "'; SET SCHEMA '" + schemaName + "'");
+
+      return sqlConnection.query(sql).execute()
+          .map(x -> {
+            connectionCache.add((PgConnection) sqlConnection);
+            return (PgConnection) sqlConnection;
+          });
+    });
+  }
+
+  private PgConnection getCachedConnection() {
+    if (connectionCache.isEmpty()) {
+      throw new IllegalStateException("No connections available");
+    }
+
+    LOG.debug("Returning cache item for tenant {} {}", currentTenant, currentIndex);
+
+    PgConnection nextConnection = connectionCache.get(currentIndex);
```

Comment:
There is no implementation to guarantee the connection is in usable state. It could be in use or even closed.

Reply:
I think this is THE question. I'm thinking the connection being actually closed isn't an issue, since the pool is open connections. Closing returns the conn to the pool, but leaves the underlying conn open. But what _could_ be an issue is that another caller grabs a connection that is in the cache and has its session pegged to whatever was set which is no longer right. But I don't think this will happen (or at least I want a mechanism to prevent it).

I think it really hinges on the lifecycle of the connections. Let's say I start two sets of SQL commands, each set that flows over their own connection, and each set having different tenant ids set. One finishes before the other. What is stopping me from giving out the connection that hasn't finished to another tenant when it is still not finished?

Reply:
I think the new implementation should resolve both of these concerns.